charms/jupyter-ui/src/charm.py

```diff
@@ -8,62 +8,51 @@
 from ops.charm import CharmBase
 from ops.main import main
 from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
-from serialized_data_interface import NoCompatibleVersions, NoVersionsListed, get_interfaces
+from serialized_data_interface import (
+    NoCompatibleVersions,
+    NoVersionsListed,
+    get_interfaces,
+)
 
 log = logging.getLogger()
 
 
-class Operator(CharmBase):
-    def __init__(self, *args):
-        super().__init__(*args)
+class CheckFailed(Exception):
+    """ Raise this exception if one of the checks in main fails. """
 
-        if not self.model.unit.is_leader():
-            log.info("Not a leader, skipping set_pod_spec")
-            self.model.unit.status = WaitingStatus("Waiting for leadership")
-            return
+    def __init__(self, msg, status=None):
```

Comment:
The argument you pass here should be `status_type`. This should look like this:
```python
class CheckFailed(Exception):
    """ Raise this exception if one of the checks in main fails. """

    def __init__(self, msg, status_type=None):
        super().__init__()

        self.msg = msg
        self.status_type = status_type
        self.status = status_type(msg)
```

Right now you have hybrid of the old and new exception and that is why everything is breaking.

Reply:
Just updated, I think it now looks like the latest version of your PRs, let me know otherwise.